### ISSUE TITLE
[AI] Q1-Q2 - Human-in-the-Loop, Evaluation, Cost Control & Local LLM

### ISSUE THREAD

**Issue by lukassherman27** [lukassherman27/Benlsey-Operating-System#353]
## Overview

This strategic issue captures the AI automation gaps identified in the comprehensive platform audit. These patterns are critical for responsible AI deployment as the platform scales.

**Dependencies:** Issue #351 (Foundation) should be addressed first.

---

### What AI Can Do Automatically vs. Needs Approval

| Automation Level | Description | BDS Example |
|-----------------|-------------|-------------|
| **Level 1: Assisted** | AI provides suggestions, human decides | Email-to-proposal linking suggestions |
| **Level 2: Partial** | AI executes sub-tasks, human reviews exceptions | Auto-categorize emails by pattern |
| **Level 3: Conditional** | AI manages end-to-end under specific conditions | Auto-link emails when confidence >97.5% |
| **Level 4: High** | AI handles most cases, human for strategic/rare | Invoice extraction from trusted vendors |

### Approval Rules by Role

| Action Category | Executive (Bill) | PM | Staff |
|----------------|-----------------|-----|-------|
| Email linking | Auto-approve high-confidence | Review suggestions | Read-only |
| Proposal creation | Auto-approve | Create with review | N/A |
| Financial data | Approve all changes | View only | N/A |
| Pattern learning | Auto-approve trusted | Suggest only | N/A |
| Client communication | Must approve | Draft only | N/A |

### Confidence Thresholds

| Confidence Level | Action | Use Case |
|-----------------|--------|----------|
| **>97.5%** | Auto-execute | Exact pattern match on learned patterns |
| **85-97.5%** | Fast-track review | Strong match, single-click approval |
| **70-85%** | Standard review | Show to user with context |
| **<70%** | Human required | Flag as "needs analysis" |

### Audit Trail Requirements

Every AI action must log:
- Timestamp, User identity
- Input data (email, document, etc.)
- AI output/decision
- Confidence score
- Model version used
- Final outcome (approved/rejected/modified)

### Schema Additions

```sql
-- Audit trail for all AI actions
CREATE TABLE ai_audit_log (
    id INTEGER PRIMARY KEY,
    timestamp DATETIME DEFAULT CURRENT_TIMESTAMP,
    action_type TEXT NOT NULL,
    user_id TEXT,
    input_summary TEXT,
    ai_output TEXT,
    confidence_score REAL,
    model_version TEXT,
    final_status TEXT, -- 'approved', 'rejected', 'modified'
    modified_output TEXT
);
```

### Deliverable
- [ ] Create `AI_AUTOMATION_GOVERNANCE.md` policy document
- [ ] Implement `ai_audit_log` table
- [ ] Add confidence threshold configuration
- [ ] Build role-based permission matrix for AI actions

**Priority: CRITICAL** - Create before expanding AI automation

---

### Core Metrics for BDS

| Metric | Formula | Target |
|--------|---------|--------|
| **Linking Accuracy** | Correct links / Total suggested links | >95% |
| **Pattern Precision** | True matches / All pattern matches | >90% |
| **Pattern Recall** | Matched emails / All matchable emails | >85% |
| **F1 Score** | 2 * (Precision * Recall) / (Precision + Recall) | >87% |
| **Answer Correctness** | Factually correct outputs / Total outputs | >95% |

### Error Types to Track
- **False Positive (FP)**: Email linked to wrong proposal/project
- **False Negative (FN)**: Email that should link but wasn't suggested
- **Hallucination**: AI suggests non-existent entity
- **Over-confidence**: High confidence but wrong answer

### Schema Additions

```sql
-- Evaluation metrics
CREATE TABLE ai_evaluation_log (
    id INTEGER PRIMARY KEY,
    timestamp DATETIME DEFAULT CURRENT_TIMESTAMP,
    suggestion_id INTEGER,
    suggestion_type TEXT, -- 'email_link', 'pattern_match', etc.
    ai_confidence REAL,
    human_verdict TEXT, -- 'correct', 'incorrect', 'partial'
    error_type TEXT, -- 'false_positive', 'false_negative', 'hallucination'
    reviewed_by TEXT,
    review_time_seconds INTEGER
);

-- A/B test results
CREATE TABLE ai_ab_tests (
    id INTEGER PRIMARY KEY,
    timestamp DATETIME DEFAULT CURRENT_TIMESTAMP,
    test_name TEXT NOT NULL,
    variant TEXT NOT NULL,
    metric_name TEXT,
    metric_value REAL,
    user_id TEXT
);
```

### Improvement Loop
```
User Action → Log Feedback → Aggregate Weekly → Analyze Patterns → Update Rules
     ↓
Email suggestion rejected → Record reason → Pattern fails 3x → Disable pattern
     ↓
Email suggestion approved → Record success → Confidence increases
```

### Metrics Dashboard

| Metric | Frequency | Alert Threshold |
|--------|-----------|-----------------|
| Suggestion acceptance rate | Daily | <80% |
| False positive rate | Weekly | >5% |
| Pattern success rate | Weekly | <90% |
| Average confidence score | Daily | <85% |
| Time to human review | Daily | >24 hours |

### Deliverable
- [ ] Create `ai_evaluation_log` table
- [ ] Build accuracy tracking dashboard
- [ ] Implement feedback collection (accept/reject tracking)
- [ ] Set up weekly accuracy reports
- [ ] Add A/B testing framework for prompts

**Priority: HIGH** - Set up before expanding AI features

---

### Budget Structure
```
Monthly AI Budget: $X
├── Email analysis: 40%
├── Pattern learning: 20%
├── Ad-hoc queries: 30%
└── Buffer: 10%
```

### Alert Thresholds
- 50%: Informational notification
- 80%: Warning to executive
- 100%: Auto-throttle non-critical requests

### Throttling by Priority

| Priority Level | Description | Throttle Behavior |
|---------------|-------------|-------------------|
| **Critical** | Client-facing, time-sensitive | Never throttle |
| **High** | Business operations (email linking) | Throttle at 90% budget |
| **Medium** | Background processing (pattern learning) | Throttle at 80% budget |
| **Low** | Analytics, reporting | Throttle at 70% budget |

### Caching Strategy
- **Response caching**: 15-30% cost reduction for repetitive queries
- **Prompt caching**: Cache stable prefixes (90% discount on cached tokens)
- **Semantic caching**: Cache similar queries, not just exact matches

Cache TTL recommendations:
- Pattern matches: 24 hours
- Project context: 1 hour
- Email analysis: No cache (unique content)

### Model Routing (Cheap vs Expensive)

| Query Type | Model | Cost/1M tokens | Use Case |
|-----------|-------|----------------|----------|
| **Simple** | Local (Ollama) | ~$0 | Pattern matching, categorization |
| **Medium** | Claude Haiku | $0.25-$1.25 | Email categorization, entity extraction |
| **Complex** | Claude Opus | $15-$75 | Novel email analysis, learning new patterns |

**Cascading approach:**
1. Try local model first
2. If confidence <80%, escalate to Haiku
3. If confidence <70%, escalate to Opus

Research shows this can reduce API costs by 40-60% without quality degradation.

### Schema Additions

```sql
-- Cost tracking
CREATE TABLE ai_cost_log (
    id INTEGER PRIMARY KEY,
    timestamp DATETIME DEFAULT CURRENT_TIMESTAMP,
    model TEXT NOT NULL,
    input_tokens INTEGER,
    output_tokens INTEGER,
    cost_usd REAL,
    feature TEXT,
    cache_hit BOOLEAN DEFAULT FALSE,
    user_id TEXT
);
```

### Deliverable
- [ ] Create `ai_cost_log` table
- [ ] Implement response caching layer
- [ ] Set up budget alerts
- [ ] Build cost monitoring dashboard
- [ ] Implement model routing logic
- [ ] Create `AI_COST_POLICY.md` document

**Priority: MEDIUM** - Implement before scaling AI usage

---

### When to Use Local vs Cloud

| Use Case | Local (Ollama) | Cloud (Claude) |
|----------|---------------|----------------|
| Pattern matching | Yes | No |
| Email categorization | Yes (fine-tuned) | Fallback |
| Entity extraction | Yes | Fallback |
| Novel email analysis | No | Yes |
| Client-facing content | No | Yes |
| Sensitive data processing | Yes (privacy) | With care |
| Offline operation | Yes | No |

### Recommended Setup

```yaml
# Development/Production
ollama:
  model: llama3.2:8b
  purpose: pattern-matching, categorization
  memory: 16GB
  gpu: Metal (Mac) or CUDA (Linux)
```

### Model Selection

| Model | Parameters | VRAM | Use Case |
|-------|-----------|------|----------|
| **Llama 3.2 3B** | 3B | 8GB | Fast categorization |
| **Llama 3.2 8B** | 8B | 12GB | General tasks |
| **Qwen 2.5 14B** | 14B | 16GB | Higher accuracy |
| **DeepSeek-Coder** | 7B | 12GB | Code/structured data |

**Recommendation:** Start with Llama 3.2 8B, benchmark against Claude Haiku, fine-tune if needed.

### Fine-Tuning Requirements
When to fine-tune:
- High-volume, repetitive task
- Domain-specific terminology (architecture, hospitality)
- Consistent accuracy issues with base model

Fine-tuning workflow:
1. Collect 500+ examples of correct email categorizations
2. Export training data from approval logs
3. Fine-tune using Ollama's Modelfile ADAPTER feature
4. Test on holdout set
5. Deploy and monitor

### Samsung TRM Concept
Samsung's Tiny Recursive Model (7M parameters) outperforms GPT-4 on reasoning tasks through recursive refinement. Monitor for specialized pattern-matching applications.

### Deliverable
- [ ] Deploy Ollama with Llama 3.2 8B
- [ ] Benchmark local vs cloud accuracy
- [ ] Create `LOCAL_LLM_POLICY.md` document
- [ ] Implement routing between local and cloud
- [ ] Fine-tune if accuracy gap exists

**Priority: LOW** - Evaluate after cloud cost patterns established

---

## Implementation Roadmap

| Phase | Focus | Duration |
|-------|-------|----------|
| **Phase 1** | Governance Foundation | Week 1-2 |
| **Phase 2** | Evaluation Infrastructure | Week 3-4 |
| **Phase 3** | Cost Control | Week 5-6 |
| **Phase 4** | Model Routing | Week 7-8 |
| **Phase 5** | Local LLM Pilot | Week 9-12 |

---

## Policy Documents to Create

1. **AI_AUTOMATION_GOVERNANCE.md**
   - Automation tiers and thresholds
   - Role-based permissions
   - Audit requirements
   - Escalation paths

2. **AI_COST_POLICY.md**
   - Monthly budget allocation
   - Alert thresholds
   - Model selection criteria
   - Caching rules
   - Cost attribution

3. **LOCAL_LLM_POLICY.md**
   - Model selection criteria
   - Deployment requirements
   - Data handling rules
   - Performance targets

---

## Success Metrics

| Metric | Current | Target (Month 3) |
|--------|---------|------------------|
| Suggestion acceptance rate | ~70% | >85% |
| False positive rate | Unknown | <5% |
| Monthly AI cost | Unknown | <$500 |
| Cache hit rate | 0% | >25% |
| Local model usage | 0% | 30-40% |

---

## Sources

- HITL: Zapier, Permit.io, UiPath, Rossum confidence thresholds
- Cost: LMSYS RouteLLM, IBM LLM Routers, Helicone caching
- Evaluation: Confident AI, Braintrust, DataCamp metrics
- Local LLM: Ollama, vLLM comparison, Samsung TRM research
- Governance: EU AI Act, FINOS AI Framework, CMU RLHF tutorial
